### PR TITLE
[Independent Agent Release] New pipeline upload step for agent package for Independent Agent Release

### PR DESCRIPTION
**Pull request by dwhyrock**
## What does this PR do?
This adds an alternate upload process for the `elastic-agent-package` pipeline for the Independent Agent Release workflow.

### Workflow
The [Independent Agent Staging pipeline](https://buildkite.com/elastic/independent-agent-release-staging) will call into the `elastic-agent-package` pipeline.

Instead of calling into the Release Manager container for uploading/processing artifacts, this will use a more simple approach to use `buildkite-agent artifact upload` to store all the artifacts, and let the calling pipeline handle the rest.

In this case, only use the new workflow when `BUILDKITE_TRIGGERED_FROM_BUILD_PIPELINE_SLUG == "independent-agent-release-staging"`

The new workflow will use the Release Manager `collect` in `dry-run` mode to verify that all of the expected artifacts are present.

### Docker Tag Adjustment
There is also an issue with Docker tags where it doesn't accept the `+` character in the tag name.  This PR also changes the `+` to a `.` if the `+` is in the version string.  We use the `.` since the `-` may be evaluated as a pre-release in semver.

### Ironbank Workaround
Currently, the `ironbank` container image does not build and use the `AGENT_PACKAGE_VERSION` env var to set the version.  This causes the Release Manager container to fail to verify since it can't find the expected `ironbank` file with the new version.

It was decided that we would not include the `ironbank` containers as part of the Independent Agent Release.  Due to that, I added a simple workaround to copy the `ironbank` container file to a new file with the proper version.

### pkg/errors Linting Fixes
There was a linting error complaining about using `github.com/pkg/errors` as an import.

The built-in `errors` import doesn't have the `Wrap` or `Wrapf` functions, so I replaced them with `fmt.Errorf()` function calls.  **This code is probably worth some scrutiny to make sure I did it correctly.**

Issue: https://github.com/elastic/ingest-dev/issues/2898

## Why is it important?
To provide a way to upload Agent Package artifacts for the Independent Agent workflow that doesn't use the Release Manager container.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Author's Checklist

- [ ]

## Related issues

- 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by AndersonQ** [comment]
@dwhyrock as it's a draft, would you mind changing the PR to draft?

File: dev-tools/mage/manifest/manifest.go
Comment (alexsapran) [inline dev-tools/mage/manifest/manifest.go:105]: nit, for parsing the version we can use the https://github.com/elastic/elastic-agent/tree/main/pkg/version package and extract the part of the version we want.

File: dev-tools/mage/dockerbuilder.go
Comment (cmacknz) [inline dev-tools/mage/dockerbuilder.go:181]: https://github.com/opencontainers/distribution-spec/issues/154

I see a dash '-' used most frequently to replace the '+', this seems to be what OCI itself was doing before https://github.com/opencontainers/distribution-spec/pull/417/files

So I have a small preference for that just to avoid doing this in a way that isn't consistent with other projects that also have this problem.

**Comment by dwhyrock** [comment]
I confirmed that this code was able to successfully upload the artifacts and set a metadata value that can be used by the triggering pipeline.

https://buildkite.com/elastic/independent-agent-release-staging/builds/52

**Comment by dwhyrock** [comment]
Example of the Independent Agent Staging pipeline calling into the `elastic-agent-package` pipeline on this branch and the new code working successfully:

https://buildkite.com/elastic/independent-agent-release-staging/builds/69

File: dev-tools/mage/dockerbuilder.go
Comment (alexsapran) [inline dev-tools/mage/dockerbuilder.go:184]: I have a question, if `b.Snapshot` is defined then the resulting tag would be 
`8.13.0.build202402191057-SNAPSHOT` is this the expected behavior for a snapshot build, or is this scenario never going to happen?

File: dev-tools/mage/manifest/manifest.go
Comment (alexsapran) [inline dev-tools/mage/manifest/manifest.go:65]: 🥇

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (alexsapran) [inline .buildkite/pipeline.elastic-agent-package.yml:155]: why we want to explicitly call the upload, we normally define the uploaded artifacts in the respected field in the `step` definition.

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (alexsapran) [inline .buildkite/pipeline.elastic-agent-package.yml:162]: How are we going to use this information later, why do we want to save them?

**Review by pazone — APPROVED** [review]

**Comment by alexsapran** [comment]
I see that this PR has a failing lint error, let's try and address those before merging. 

It might seem strange, but in this way, we strive towards a more consistent codebase and potentially find and fix issues.

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (dwhyrock) [reply .buildkite/pipeline.elastic-agent-package.yml:162]: By setting the `triggered_build_id` metadata field in the build that triggered this one, it passes is the Build ID of this `elastic-agent-package` build to the triggering build.  

The triggering build then uses this in a command to pull down the artifacts from this specific build.

Here's the command in the `unified-release` repo where we reference this: https://github.com/elastic/unified-release/blob/main/catalog/step-get-agent-bk-artifacts.go#L15

It's currently one of the easiest ways for a calling build to pull artifacts back from a triggered build in Buildkite.  Other ways involve obtaining a Buildkite API token that then has to be managed in vault, etc.  

Here's the example we took this from: https://forum.buildkite.community/t/how-to-download-artifacts-back-from-triggered-pipeline/3480/2

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (dwhyrock) [reply .buildkite/pipeline.elastic-agent-package.yml:155]: My thought was that we are verifying that the artifacts are all present with the above `--dry-run` of the `dra-publish.sh` script.  The normal DRA step here relies on the Release Manager container to do the proper uploads, so the expected files aren't currently defined in this repo.

We will have code for the Independent Release added to the `unified-release` repo that will verify that each expected file is present, which will happen after the artifacts are downloaded from this build.

If we also do the expected file check here, and if there's a change in which artifacts are expected, then we will have two places to change it instead of one.

With the Buildkite migration of the main Unified Release workflow, I believe there will be something that eventually replaces the Release Manager container to upload/verify the artifacts.  But that's more on David and Alpar's plate since I'm not on the Release Eng team.

File: dev-tools/mage/dockerbuilder.go
Comment (dwhyrock) [reply dev-tools/mage/dockerbuilder.go:184]: We only plan on having a Staging build for the Independent Agent Release.  No Snapshot builds.

**Comment by dwhyrock** [comment]
> I see that this PR has a failing lint error, let's try and address those before merging.

I will work on fixing the linting errors.

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (alexsapran) [reply .buildkite/pipeline.elastic-agent-package.yml:162]: oh ok, it make sense then. 

Can you put some comment about this reference, because if someone is going to touch this line and break it, the time to debug would be huge, especially if they don't have the full context.

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (alexsapran) [reply .buildkite/pipeline.elastic-agent-package.yml:155]: Why we need explicitly to call the `buildkite-agent artifact upload` is my point. 
We normally set it like this
```yaml
  - label: "Blah Blah"
    command: run_me.sh
    artifact_paths:
      - "build/distributions/*"
```

This way no matter your code inside the step, at the end, buildkite is going to try and upload those artifacts.

File: dev-tools/mage/dockerbuilder.go
Comment (dwhyrock) [inline dev-tools/mage/dockerbuilder.go:49]: It's probably worth someone to look over these changes from `pkg/errors` (due to linting complaints) to `fmt.Errorf` calls to make sure I did them correctly.

File: .buildkite/pipeline.elastic-agent-package.yml
Comment (dwhyrock) [reply .buildkite/pipeline.elastic-agent-package.yml:155]: I will change it to use the `artifact_paths` entry instead of `buildkite-agent artifact upload`

**Comment by dwhyrock** [comment]
I believe I've addressed all feedback and this is ready for another review.

**Review by pierrehilbert — APPROVED** [review]
Thx @dwhyrock for your work here!

**Pull request merged**
